File: src/nodes/cms_generator.py
import re
import json
from typing import Dict, Any, List
from langchain_core.messages import SystemMessage, HumanMessage
from src.state import GlobalState, SectionInfo
from src.llm import get_llm
from src.prompts.cms_generator_prompt import CMS_GENERATOR_SYSTEM_PROMPT
from src.utils.token_tracker import log_token_usage
from src.nodes.section_identifier import extract_content_text


def parse_json_block(response_text: str) -> dict:
    """
    Extracts and parses a JSON code block from the model's response.
    """
    # Try to find ```json ... ``` block
    json_match = re.search(r'```json\s*\n(.*?)```', response_text, re.DOTALL)
    if json_match:
        content = json_match.group(1).strip()
    else:
        # Fallback: check generic code block
        code_match = re.search(r'```\s*\n(.*?)```', response_text, re.DOTALL)
        if code_match:
            content = code_match.group(1).strip()
        else:
            # Last fallback: clean text
            content = response_text.strip()

    # Clean up JSON formatting artifacts if present
    content = content.replace("```json", "").replace("```", "").strip()
    
    return json.loads(content)


def generate_cms_for_single_section(
    section_info: dict,
    previous_sections: List[dict],
    project_name: str,
    page_name: str = "home"
) -> dict:
    """
    Generates the CMS JSON schema for a single section.
    """
    section_id = section_info["id"]
    section_name = section_info["name"]
    section_description = section_info["description"]
    
    print(f"  [CMS Generator] Generating CMS configuration for section {section_id}: '{section_name}'...")
    
    # 1. Build context of previous sections to avoid name/index collisions
    prev_summary_items = []
    for prev in previous_sections:
        if prev.get("cms"):
            meta = prev["cms"].get("metadata", {})
            elements = prev["cms"].get("elements", [])
            elem_names = [e.get("elementName") for e in elements if e.get("elementName")]
            prev_summary_items.append(
                f"- Section ID {meta.get('sectionId')} ('{meta.get('sectionName')}'): "
                f"Index={meta.get('index')}, ElementNames={elem_names}"
            )
            
    if not prev_summary_items:
        previous_cms_summary = "None (this is the first section)."
    else:
        previous_cms_summary = "\n".join(prev_summary_items)

    # 2. Initialize Gemini LLM (using low temperature for strict schema adherence)
    llm = get_llm(temperature=0.2)
    
    # 3. Format the prompts
    system_prompt = CMS_GENERATOR_SYSTEM_PROMPT.format(
        project_name=project_name,
        page_name=page_name,
        previous_cms_summary=previous_cms_summary,
        section_name=section_name,
        section_description=section_description
    )
    
    messages = [
        SystemMessage(content=system_prompt),
        HumanMessage(content="Generate the CMS JSON matching the exact schema requirements.")
    ]
    
    try:
        response = llm.invoke(messages)
        log_token_usage(f"CMS Generator ({section_name})", response)
        
        response_text = extract_content_text(response.content)
        cms_dict = parse_json_block(response_text)
        
        # 4. Strict validation and formatting checks
        metadata = cms_dict.get("metadata", {})
        elements = cms_dict.get("elements", [])
        
        # Ensure metadata values exist
        metadata["sectionName"] = section_name
        metadata["pageName"] = page_name
        
        # Ensure section index is correct and sequential
        planned_index = section_id # Default sequential order matching section planning ID
        if previous_sections:
            prev_indexes = [
                p["cms"]["metadata"].get("index", 0) 
                for p in previous_sections if p.get("cms") and "metadata" in p["cms"]
            ]
            max_prev = max(prev_indexes) if prev_indexes else 0
            metadata["index"] = max(max_prev + 1, planned_index)
        else:
            metadata["index"] = planned_index
            
        # propagate generated section ID to all elements
        import uuid
        section_id_val = str(uuid.uuid4())
        metadata["sectionId"] = section_id_val
        
        for elem in elements:
            elem["sectionId"] = section_id_val
            elem["section"] = section_name
            elem["projectName"] = project_name
            elem["pageName"] = page_name
            
            # Ensure globally unique fieldId
            elem["fieldId"] = str(uuid.uuid4())
            
            # Ensure globally unique fieldId inside loop arrays
            if "loop" in elem and isinstance(elem["loop"], list):
                for item in elem["loop"]:
                    for i in range(1, 11):
                        field_id_key = f"fieldId{i}"
                        if field_id_key in item:
                            item[field_id_key] = str(uuid.uuid4())
            
        cms_dict["metadata"] = metadata
        cms_dict["elements"] = elements
        
        print(f"  [CMS Generator] ✓ CMS generated successfully for section {section_id}")
        return cms_dict
        
    except Exception as e:
        error_msg = str(e)
        print(f"  [CMS Generator] ✗ CMS generation FAILED for section {section_id}: {error_msg}")
        # Fallback baseline CMS schema if model fails to parse or generate
        import uuid
        fallback_sec_id = str(uuid.uuid4())
        fallback_cms = {
            "metadata": {
                "sectionId": fallback_sec_id,
                "sectionName": section_name,
                "sectionStatus": "Active",
                "variations": 1,
                "sectionType": "",
                "path": f"/client/{project_name}/{section_name.replace(' ', '')}/Variation1",
                "isAiGenerated": True,
                "pageName": page_name,
                "index": section_id
            },
            "elements": [
                {
                    "sectionId": fallback_sec_id,
                    "elementName": f"{section_name.lower().replace(' ', '')}Title",
                    "fieldId": str(uuid.uuid4()),
                    "content": section_name,
                    "contentType": "Text",
                    "section": section_name,
                    "projectName": project_name,
                    "pageName": page_name,
                    "isCustom": True,
                    "isCustomEdit": True
                }
            ]
        }
        return fallback_cms

File: src/nodes/cms_generator.ts
import crypto from "node:crypto";
import { SystemMessage, HumanMessage } from "@langchain/core/messages";
import { getLlm } from "../llm.js";
import { CMS_GENERATOR_SYSTEM_PROMPT } from "../prompts/cms_generator_prompt.js";
import { logTokenUsage } from "../utils/token_tracker.js";

function extractContentText(content: any): string {
  if (typeof content === "string") {
    return content.trim();
  }
  if (Array.isArray(content)) {
    return content
      .map((part) => (typeof part === "string" ? part : part.text || ""))
      .join("")
      .trim();
  }
  return String(content).trim();
}

function parseJsonBlock(responseText: string): any {
  const jsonMatch = responseText.match(/```json\s*\n([\s\S]*?)```/);
  let content = "";
  if (jsonMatch) {
    content = jsonMatch[1].trim();
  } else {
    const codeMatch = responseText.match(/```\s*\n([\s\S]*?)```/);
    if (codeMatch) {
      content = codeMatch[1].trim();
    } else {
      content = responseText.trim();
    }
  }

  content = content.replace("```json", "").replace("```", "").trim();
  return JSON.parse(content);
}

/**
 * Generates the CMS JSON schema for a single section.
 */
export async function generateCmsForSingleSection(
  sectionInfo: { id: number; name: string; description: string },
  previousSections: any[],
  projectName: string,
  pageName: string = "home"
): Promise<any> {
  const sectionId = sectionInfo.id;
  const sectionName = sectionInfo.name;
  const sectionDescription = sectionInfo.description;

  console.log(`  [CMS Generator] Generating CMS configuration for section ${sectionId}: '${sectionName}'...`);

  const prevSummaryItems: string[] = [];
  for (const prev of previousSections) {
    if (prev && prev.cms) {
      const meta = prev.cms.metadata || {};
      const elements = prev.cms.elements || [];
      const elemNames = elements.map((e: any) => e.elementName).filter(Boolean);
      prevSummaryItems.push(
        `- Section ID ${meta.sectionId} ('${meta.sectionName}'): Index=${meta.index}, ElementNames=${elemNames}`
      );
    }
  }

  const previousCmsSummary =
    prevSummaryItems.length === 0 ? "None (this is the first section)." : prevSummaryItems.join("\n");

  const llm = getLlm(0.2);

  const systemPrompt = CMS_GENERATOR_SYSTEM_PROMPT
    .replace("{project_name}", projectName)
    .replace("{page_name}", pageName)
    .replace("{previous_cms_summary}", previousCmsSummary)
    .replace("{section_name}", sectionName)
    .replace("{section_description}", sectionDescription)
    .replace("{index}", String(sectionId)); // Simple replacement if template string format is used

  const messages = [
    new SystemMessage(systemPrompt),
    new HumanMessage("Generate the CMS JSON matching the exact schema requirements."),
  ];

  try {
    const response = await llm.invoke(messages);
    logTokenUsage(`CMS Generator (${sectionName})`, response);

    const responseText = extractContentText(response.content);
    const cmsDict = parseJsonBlock(responseText);

    const metadata = cmsDict.metadata || {};
    const elements = cmsDict.elements || [];

    metadata.sectionName = sectionName;
    metadata.pageName = pageName;

    const plannedIndex = sectionId;
    if (previousSections && previousSections.length > 0) {
      const prevIndexes = previousSections
        .map((p) => p?.cms?.metadata?.index)
        .filter((idx) => typeof idx === "number");
      const maxPrev = prevIndexes.length > 0 ? Math.max(...prevIndexes) : 0;
      metadata.index = Math.max(maxPrev + 1, plannedIndex);
    } else {
      metadata.index = plannedIndex;
    }

    const sectionIdVal = crypto.randomUUID();
    metadata.sectionId = sectionIdVal;

    for (const elem of elements) {
      elem.sectionId = sectionIdVal;
      elem.section = sectionName;
      elem.projectName = projectName;
      elem.pageName = pageName;
      elem.fieldId = crypto.randomUUID();

      if (elem.loop && Array.isArray(elem.loop)) {
        for (const item of elem.loop) {
          for (let i = 1; i <= 10; i++) {
            const fieldIdKey = `fieldId${i}`;
            if (fieldIdKey in item) {
              item[fieldIdKey] = crypto.randomUUID();
            }
          }
        }
      }
    }

    cmsDict.metadata = metadata;
    cmsDict.elements = elements;

    console.log(`  [CMS Generator] ✓ CMS generated successfully for section ${sectionId}`);
    return cmsDict;
  } catch (error: any) {
    console.error(`  [CMS Generator] ✗ CMS generation FAILED for section ${sectionId}: ${error.message}`);
    const fallbackSecId = crypto.randomUUID();
    return {
      metadata: {
        sectionId: fallbackSecId,
        sectionName: sectionName,
        sectionStatus: "Active",
        variations: 1,
        sectionType: "",
        path: `/client/${projectName}/${sectionName.replace(/\s+/g, "")}/Variation1`,
        isAiGenerated: true,
        pageName: pageName,
        index: sectionId,
      },
      elements: [
        {
          sectionId: fallbackSecId,
          elementName: `${sectionName.toLowerCase().replace(/\s+/g, "")}Title`,
          fieldId: crypto.randomUUID(),
          content: sectionName,
          contentType: "Text",
          section: sectionName,
          projectName: projectName,
          pageName: pageName,
          isCustom: true,
          isCustomEdit: true,
        },
      ],
    };
  }
}
